lib/galaxy/managers/datasets.py

```diff
@@ -686,15 +686,16 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
         return unwrapped_val
 
     def deserialize_datatype(self, item, key, val, **context):
-        if item.datatype.allow_datatype_change and self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
-            if DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
-                item.change_datatype(val)
-                context['trans'].sa_session.flush()
-                self.app.datatypes_registry.set_external_metadata_tool.tool_action.execute(self.app.datatypes_registry.set_external_metadata_tool, context['trans'], incoming={'input1': item}, overwrite=False)  # overwrite is False as per existing behavior
-                return item.datatype
-        context['trans'].response.status = 500
-        context['trans'].error_message = 'Datatype could not be updated.'
-        raise Exception('Datatype could not be updated.')
+        if not item.datatype.allow_datatype_change:
+            raise Exception("The current datatype does not allow datatype changes.")
```

Comment:
You can use a more specific exception (need a `from galaxy.exceptions import RequestParameterInvalidException` )

```suggestion
            raise RequestParameterInvalidException("The current datatype does not allow datatype changes.")
```